lib/cn/kvset_split.c

```diff
@@ -0,0 +1,619 @@
+                         num_ptombs, hlog_data(work->hlog_left), ptree, ptree_pgc);
+        if (err)
+            return err;
+
+        result->blks_left.hblk = hblk;
+        blk_list_append(&result->blks_commit, hblk.bk_blkid);
+    }
+
+    if (result->blks_right.kblks.n_blks > 0 || ptree_pgc > 0) {
+        err = hbb_finish(work->hbb_right, &hblk, result->vgmap_right,
+                         &min_pfx, &max_pfx, min_seqno, max_seqno,
+                         result->blks_right.kblks.n_blks, result->blks_right.vblks.n_blks,
+                         num_ptombs, hlog_data(work->hlog_right), ptree, ptree_pgc);
+        if (err)
+            return err;
+
+        result->blks_right.hblk = hblk;
+        blk_list_append(&result->blks_commit, hblk.bk_blkid);
+    }
+
+    blk_list_append(&result->blks_purge, ks->ks_hblk.kh_hblk_desc.mbid);
+
+    return 0;
+}
+
+static void
+kvset_split_free(
+    struct kvset            *ks,
+    struct kvset_split_work *work,
+    struct kvset_split_res  *result,
+    bool                     del_mblks)
+{
+    hbb_destroy(work->hbb_left);
+    hbb_destroy(work->hbb_right);
+
+    hlog_destroy(work->hlog_left);
+    hlog_destroy(work->hlog_right);
+
+    blk_list_free(&result->blks_left.kblks);
+    blk_list_free(&result->blks_right.kblks);
+
+    blk_list_free(&result->blks_left.vblks);
+    blk_list_free(&result->blks_right.vblks);
+
+    blk_list_free(&result->blks_purge);
+
+    for (int i = 0; del_mblks && i < result->blks_commit.n_blks; i++)
+        delete_mblock(ks->ks_mp, &result->blks_commit.blks[i]);
+    blk_list_free(&result->blks_commit);
+
+    kvset_vgmap_free(result->vgmap_left);
+    kvset_vgmap_free(result->vgmap_right);
+}
+
+static merr_t
+kvset_split_alloc(
+    struct kvset            *ks,
+    struct kvset_split_work *work,
+    struct kvset_split_res  *result)
+{
+    struct cn *cn = cn_tree_get_cn(ks->ks_tree);
+    merr_t err = 0;
+    uint32_t nvgroups = kvset_get_vgroups(ks);
+
+    memset(result, 0, sizeof(*result));
+
+    result->vgmap_left = kvset_vgmap_alloc(nvgroups);
+    result->vgmap_right = kvset_vgmap_alloc(nvgroups);
+    if (!result->vgmap_left || !result->vgmap_right) {
+        err = merr(ENOMEM);
+        goto errout;
+    }
+
+    err = hbb_create(&work->hbb_left, cn);
+    if (err)
+        goto errout;
+
+    err = hbb_create(&work->hbb_right, cn);
+    if (err)
+        goto errout;
+
+    err = hlog_create(&work->hlog_left, HLOG_PRECISION);
+    if (err)
+        goto errout;
+
+    err = hlog_create(&work->hlog_right, HLOG_PRECISION);
+    if (err)
+        goto errout;
+
+    blk_list_init(&result->blks_left.kblks);
+    blk_list_init(&result->blks_right.kblks);
+
+    blk_list_init(&result->blks_left.vblks);
+    blk_list_init(&result->blks_right.vblks);
+
+    blk_list_init(&result->blks_purge);
+    blk_list_init(&result->blks_commit);
+
+errout:
+    if (err)
+        kvset_split_free(ks, work, result, false);
```

Comment:
Are we not leaking memory here in the case of an error? Creating hlogs and hbbs. Not entirely a fan of the pattern of relying on the caller to free memory in the case of an error.

Reply:
kvset_split_free takes care of destroying the hlogs and hbbs in the error path.